check that initial guesses are within bounds. Fixes #355

Peter230655 opened the pull request:
A method, called from solve to ensure that the initial guesses are within their respective bounds. As such a conflict is not fatal for opty, a UserWarning is raised. Should take care of issue #355

moorepants commented on opty/direct_collocation.py:
```diff
@@ -249,9 +250,57 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
 
         """
-
+        self.bounds_conflict_initial_guess(free)
```
After some thought, I think I prefer that you have:

```python
        if not allow_out_of_bounds:
            self.bounds_conflict_initial_guess(free)
```

and it shoudl be a kwarg of `.solve()` that is set to false by default. Then an error is raised. The user may have a ligitmate reason to not want this, so they can disable it with `allow_out_of_bounds=True`.

Feel free to suggest a better name for the kwarg and/or use the "not" of it, i.e. `error_out_of_bounds` or something. I haven't thought of a shorter name for the `kwarg`.

There is a tendency to avoid use of warnings because they are easy to ignore or may not display at all depending on the settings Python is run with. So forcing an error is usually the right choice.

Peter230655 replied:
> After some thought, I think I prefer that you have:
> 
> ```python
>         if not allow_out_of_bounds:
>             self.bounds_conflict_initial_guess(free)
> ```
> 
> and it shoudl be a kwarg of `.solve()` that is set to false by default. Then an error is raised. The user may have a ligitmate reason to not want this, so they can disable it with `allow_out_of_bounds=True`.
> 
> Feel free to suggest a better name for the kwarg and/or use the "not" of it, i.e. `error_out_of_bounds` or something. I haven't thought of a shorter name for the `kwarg`.
> 
> There is a tendency to avoid use of warnings because they are easy to ignore or may not display at all depending on the settings Python is run with. So forcing an error is usually the right choice.

What about ``respect_bounds``? Default would be True, and an error raised. What error should be raised? ``ValueError``?
Just so I get it right: Some old code, where the bounds are ignored by the initial guess would suddenly get an error. Tihs does not matter?

moorepants replied:
Yes, much better name!

ValueError is appropriate.

>  Just so I get it right: Some old code, where the bounds are ignored by the initial guess would suddenly get an error. This does not matter?

We'll have to update any examples or tests to accommodate this. Either by changing initial guess (probably best idea) or setting `respect_bounds=False`.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1908,3 +1908,151 @@ def test_attributes_read_only():
+    start_idx = np.random.randint(0, num_nodes - 1)
+    for i in range(9):
+        initial_guess[start_idx + i*num_nodes] = -10.0
+    initial_guess[-1] = 0.5
```
You need a test that calls the new method explicitly as the primary test, i.e. does your new method do what it is supposed to do. This doesn't not require solving the problem.

Then additionally you need the test that `.solve()` does the right thing with and without the suggested flag above.

moorepants replied:
Once you have the new `kwarg` you need to test if `True` and `False` do the right things.

Peter230655 commented:
Adjusted as per the suggestions given.

Peter230655 commented:
If I read it correctly, about 10 examples seem to give problems.
What I could do is this:
- set the kwarg to False
- over the next so many days, I could look at the offending examples, and set their kwarg to False?
Or what would you suggest?

moorepants commented:
My suggestion was made above:

> We'll have to update any examples or tests to accommodate this. Either by changing initial guess (probably best idea) or setting respect_bounds=False.

moorepants commented:
Good looking error message!

>     ValueError: The initial guesses for [c, k] are in conflict with their bounds.

Peter230655 commented:
> Good looking error message!
> 
> > ```
> > ValueError: The initial guesses for [c, k] are in conflict with their bounds.
> > ```

Thanks!
I wasn't sure whether these brackets [....] would be o.k. - but frankly I could not remove them.

Peter230655 commented:
> My suggestion was made above:
> 
> > We'll have to update any examples or tests to accommodate this. Either by changing initial guess (probably best idea) or setting respect_bounds=False.

I will go over the examples in the coming days, try to fix them and push them if fixed.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -204,7 +204,7 @@ def __init__(self, obj, obj_grad, equations_of_motion, state_symbols,
         self.obj_value = []
 
-    def solve(self, free, lagrange=[], zl=[], zu=[]):
+    def solve(self, free, respect_bounds=True, lagrange=[], zl=[], zu=[]):
```
```suggestion
    def solve(self, free,lagrange=[], zl=[], zu=[],  respect_bounds=True):
```

Needs to go at end to not break backwards compatibility.

Peter230655 replied:
> Needs to go at end to not break backwards compatibility.

I did not know this! I thought with kwargs the sequence did not matter.

moorepants replied:
Technically it does because this works: `prob.solve(free, my_lagrange, my_zl, my_zu`. If you don't specify the kwarg name when calling the function/method, it uses the order. Some people may do this.

That's why Python introduced a new syntax, something like `def solve(self, free, *, lagrange=[], ...)` that actually enforces you to name the kwarg on use (I think). But if we add that, it is backwards incompatible.

Peter230655 replied:
> Technically it does because this works: `prob.solve(free, my_lagrange, my_zl, my_zu`. If you don't specify the kwarg name when calling the function/method, it uses the order. Some people may do this.
> 
> That's why Python introduced a new syntax, something like `def solve(self, free, *, lagrange=[], ...)` that actually enforces you to name the kwarg on use (I think). But if we add that, it is backwards incompatible.

Thanks! My python knowledge still so limited...

moorepants commented on opty/direct_collocation.py:
```diff
@@ -249,9 +254,58 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
+        """
+        Ascertains that the initial guesses for all variables are within the
+        limits prescribed by their respective bounds. As such possible
+        conflicts are not fatal for opty, only a UserWarning is raised.
```
Say "raises a ValueError"

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1908,3 +1908,243 @@ def test_attributes_read_only():
+    with raises(ValueError):
+        prob.bounds_conflict_initial_guess(initial_guess)
+
+def test_solve_kwarg_respect_bounds():
```
My guess is there is no reason to have two test functions. Can this not all be in one?

Peter230655 replied:
> My guess is there is no reason to have two test functions. Can this not all be in one?

Can easily be in one. I thought two were wanted.

moorepants replied:
You have lots of set up code often in these, so it is best to minimize that in tests, especially if it is compiling the C code which always takes some time.

Peter230655 replied:
> You have lots of set up code often in these, so it is best to minimize that in tests, especially if it is compiling the C code which always takes some time.

I fully agree. When I looked at the running time, most of it was this 'set up'.
So, while normally different things should have different test function, but if the different things are closely related, like here, one test function is acceptable?

moorepants replied:
I often test many things in one test function, especially if setup is computationally costly.

Peter230655 commented:
I proceeded as follows:
- Where possible, I adjusted the initial guesses to meet the bounds.
-  where solutions of previous runs were used as initial guesses, I opened the bounds slightly to allow for numerical errors in the solutions.
- ``plot_park2004.py``: Here the initial guesses for the gain factors ``k00``.... ``k13`` were very far from their bounds. When I set them to meet the bounds, the results seemed to be a bit worse. So here I set ``recpect_bounds=False``
- ``plot_human_trait.py``. Here the initial guess for the earth solution was some previous solution. Its solution is used as the initial guess for the moon solution. THe earth solution seemed far outside the bounds. So, I used the initial guess for the earth solution also for the moon solution

Peter230655 commented:
This test has been running for nearly 2 1/2 hours.
Is this normal, or did something happen?

Peter230655 commented:
seems to run too long, will start again.

moorepants commented:
it seems that conda cannot resolve the environment on windows with python 3.10.

```
yeadon                    1.5.0              pyhd8ed1ab_2    conda-forge
zstandard                 0.23.0          py310he5e10e1_1    conda-forge
zstd                      1.5.6                h0ea2cb4_0    conda-forge
CI detected...
Channels:
 - conda-forge
 - defaults
Platform: win-64
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... 
Error: The operation was canceled.
```

I don't know why that would be, as it has worked before.

moorepants commented:
Peter, since you are on windows can you make this edit to your conda env file and see if it works locally when creating the environment:

```
name: opty-dev
channels:
  - conda-forge
dependencies:
  # build
  - setuptools
  # run (includes optional run deps)
  - cyipopt >=1.1.0
  - cython >=0.29.19
  - matplotlib >=3.2.0
  - numpy >=1.19.0
  - python =3.10
  - scipy >=1.5.0
  - sympy >=1.13  # version required for the docs
  # dev
  - coverage
  - ipython
  - pytest
  - pytest-cov
  # docs/examples
  - joblib
  - numpydoc
  - pip <24  # pin required for the gait2d pip install below
  - pydy >=0.5.0  # gait2d
  - pyyaml  # gait2d
  - sphinx
  - sphinx-gallery
  - sphinx-reredirects
  - symmeplot
  - yeadon
  - pip:
    - -e git+https://github.com/csu-hmc/gait2d#egg=gait2d
```

Peter230655 commented:
> Peter, since you are on windows can you make this edit to your conda env file and see if it works locally when creating the environment:
> 
> ```
> name: opty-dev
> channels:
>   - conda-forge
> dependencies:
>   # build
>   - setuptools
>   # run (includes optional run deps)
>   - cyipopt >=1.1.0
>   - cython >=0.29.19
>   - matplotlib >=3.2.0
>   - numpy >=1.19.0
>   - python =3.10
>   - scipy >=1.5.0
>   - sympy >=1.13  # version required for the docs
>   # dev
>   - coverage
>   - ipython
>   - pytest
>   - pytest-cov
>   # docs/examples
>   - joblib
>   - numpydoc
>   - pip <24  # pin required for the gait2d pip install below
>   - pydy >=0.5.0  # gait2d
>   - pyyaml  # gait2d
>   - sphinx
>   - sphinx-gallery
>   - sphinx-reredirects
>   - symmeplot
>   - yeadon
>   - pip:
>     - -e git+https://github.com/csu-hmc/gait2d#egg=gait2d
> ```

So I understand correctly:
- the above is a .yml file.
- I store it somewhere on my PC
- then I create a new environment, called opty-dev, using this .yml file like it is described in opty, 
- OR would I use this .yml file to update my existing opty-dev environment? This I do not know how to do.

Thanks!

moorepants commented:
You can do either, but I only changed this line `  - python =3.10` so it would install python 3.10.

Peter230655 commented:
> You can do either, but I only changed this line ` - python =3.10` so it would install python 3.10.

So to update I proceed as follows?
1. I take the yml file above and store it on my PC as, say, ``update.yml`` 
2. I open the anaconda promt and put ``activate opty-dev``
3. then I put ``conda env update -f path-to-update.yml``

moorepants commented:
That will work but only if you first remove the existing opty-dev conda environment.

moorepants commented:
Sorry, nevermind, if you use "update" then it should ve fine. You can also simple add `python =3.10` to the yml file you already have in the repo.

Peter230655 commented:
> Sorry, nevermind, if you use "update" then it should be fine. You can also simple add `python =3.10` to the yml file you already have in the repo.

This looks easiest for me! Then, when I push, this changed .yml file will also be pushed. Is this o.k.?

Peter230655 commented:
Unsure, if this is related to the issue:
When I run ``plot_sit_to_stand`` (This file had not problems with the bounds so I did not touch it!) I get very poorly met eoms:
![image](https://github.com/user-attachments/assets/0d4126e5-cc20-4c9c-9ef8-3f33aa3d8ae1)
I also get this error:
![image](https://github.com/user-attachments/assets/9a43e054-a433-4a0e-9334-983a19d216a7)

moorepants commented:
No you should not push this change to the yml file. We do not want to specify the python version in the yml file stored in this repo.

moorepants commented:
After seeing how much you have changed to get this to work, I'm second guessing my advice to make `respect_bounds=True` the default. This is a backwards incompatible change and will cause a lot of code to break. Maybe we should make the default be False.

Peter230655 commented:
> After seeing how much you have changed to get this to work, I'm second guessing my advice to make `respect_bounds=True` the default. This is a backwards incompatible change and will cause a lot of code to break. Maybe we should make the default be False.

This backwards incompatability was a slight concern I had with setting ``respect_bounds=True``, so I agree with your idea completely.
Maybe best to close this PR and make a new one where I only change ``direct_collocation.py`` and add the test to ``test_direct_collocation.py``?

moorepants commented:
I'd try to get out of the habit of closing and opening new prs, you can simply update this one until it is where we like it.

Peter230655 commented:
> I'd try to get out of the habit of closing and opening new prs, you can simply update this one until it is where we like it.

I will do so, within today, I hope.

moorepants commented:
If you have trouble with conda environments, remember that I wrote this guide: https://mechmotum.github.io/blog/setup-development-environment.html

moorepants commented:
I opened an issue about the build failure: https://github.com/conda-forge/conda-forge.github.io/issues/2460

Peter230655 commented:
> I opened an issue about the build failure: [conda-forge/conda-forge.github.io#2460](https://github.com/conda-forge/conda-forge.github.io/issues/2460)

I am switching to ``respect_bounds=False`` as default, and change the examples back to their original state. Then I will push again. Let's see.

moorepants commented:
The Windows + Python 3.10 error is unrelated to your changes, so it will likely continue to fail. If you can try to install the environment on your computer with Python 3.10, that may give us more clues (as you can see the whole error message) or let it run longer.

Peter230655 commented:
> The Windows + Python 3.10 error is unrelated to your changes, so it will likely continue to fail. If you can try to install the environment on your computer with Python 3.10, that may give us more clues (as you can see the whole error message) or let it run longer.

If I install Python 3.10 in my environment opty-dev, and then run it on my PC the error should show, correct?

Peter230655 commented:
I may have found an error:
I switched my opty-dev environment to python 3.10
When I ran the examples I got an error which I did not get before. I had;
``print(f'some text {info['obj_val']} more text')`` This gave an error. Then I switched to
``print(f"some text {info['obj_val']} more text")`` and it ran fine. Apparently an issue with python 3.10

Peter230655 commented:
Since I found an error which seems to happen only with python 3.10 I risk a push again.

moorepants commented:
The error is that the python 3.10 environment on windows will not install (as far as I can tell). If you successfully installed it locally, that is important info. How long did it take to solve?

Peter230655 commented:
> The error is that the python 3.10 environment on windows will not install (as far as I can tell). If you successfully installed it locally, that is important info. How long did it take to solve?

What I did was this:
- I copied your .yml file to some other directory
- I set python =3.10
- then I updated my opty-dev with this .yml file. 
- when I made conda list it showed python 3.10.xx (I do not recall the xx number)

then I ran to create the html files. This took maybe 5 min.

Peter230655 commented:
> If you have trouble with conda environments, remember that I wrote this guide: https://mechmotum.github.io/blog/setup-development-environment.html

I printed this the first time you sent to me! :-) ...... but I have this tendency to forget. So thanks for telling me again!

Peter230655 commented:
Should I switch my opty-dev to Python 3.13 to see what has happened now?

moorepants commented:
> This took maybe 5 min to run.

If it took 5 minutes to make the environment, that is close to the 360 second time limit on CI. You have a newer computer, so it may be that the solve is just taking too much time for that. I think the only solution is to narrow the version qualifiers in the yaml file to reduce the possible combinations of software.

Peter230655 commented:
> > This took maybe 5 min to run.
> 
> If it took 5 minutes to make the environment, that is close to the 360 second time limit on CI. You have a newer computer, so it may be that the solve is just taking too much time for that. I think the only solution is to narrow the version qualifiers in the yaml file to reduce the possible combinations of software.

Sorry, I meant 5 min to make the html files. Updating my environment to python 3.10 was maybe one min.

moorepants commented:
Can you create the environment from scratch, not simply update. It is failing to create the environment from scratch in CI.

moorepants commented:
This one has been running for 5 hours and 30 minutes and hasn't solved the environment: https://github.com/csu-hmc/opty/actions/runs/13353531780/job/37293136717?pr=361

Peter230655 commented:
> Can you create the environment from scratch, not simply update. It is failing to create the environment from scratch in CI.

NO problem! I will use the .yml file in opty, right?

Peter230655 commented:
I built from scratch as follows:
- I deleted the environment opty-dev
- I created a new one as explained in the opty documentation. 
This took about 90 dec

moorepants commented:
>  Should I do the same thing but change line 12 there from python to python =3.10 ?

Yes, you have to set python = 3.10 to do the same thing in the failing CI build.

Peter230655 commented:
> > Should I do the same thing but change line 12 there from python to python =3.10 ?
> 
> Yes, you have to set python = 3.10 to do the same thing in the failing CI build.

Clear, will do so. Right now I managed once again to mess up **everything** :-((
Hopefully I'll fix it tomorrow morning, and report about the timing.

moorepants commented:
If you merge master the tests should run correctly again. I fixed it in #363

Peter230655 commented:
Worked !!   :-)

moorepants commented on opty/direct_collocation.py:
```diff
@@ -244,9 +249,58 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
+        ----------
+        free : array-like, shape(n*N + q*N + r + s, )
+            Initial guess given to solve.
+
```
You can add a "Raises" section to describe the ValueError. See the numpydoc documentation.

Peter230655 replied:
> You can add a "Raises" section to describe the ValueError. See the numpydoc documentation.

I assume you meant this section:
![image](https://github.com/user-attachments/assets/c214e6f8-0966-484e-b766-3e6c7fb5ac8e)

moorepants commented on opty/direct_collocation.py:
```diff
@@ -244,9 +249,58 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
+                        violating_variables.append(local_ts)
+
+            symbole = self.collocator.state_symbols + \
+            self.collocator.unknown_input_trajectories
```
bit better to do:

```
a = (longgggggggggggggggggggggg +
     longgggggggggggggggggggggggggg)
```

Peter230655 replied:
> bit better to do:
> 
> ```
> a = (longgggggggggggggggggggggg +
>      longgggggggggggggggggggggggggg)
> ```

So, parenthesis are always preferred over \ ?

moorepants replied:
according to pep 8 it is

Peter230655 replied:
Thanks!
I would have thought the other way around, I will remember in future.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -244,9 +249,58 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
+
+            if len(violating_variables) > 0:
+                msg = f'The initial guesses for {violating_variables} are in '+\
+                f'conflict with their bounds.'
```
strings don't need `\` if you do this:

```python
msg = (f'sstringgggggggggggggggggggggggg'
       f'anotherstringgggggg')
```

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,161 @@ def test_attributes_read_only():
+    par_map = {b1: 1.0,
+               b2: 2.0,
+               b3: 3.0
+               }
```
preferred formating for dictionaries is:

```python
a = {
    'b': 1,
    'c': 2,
}
```

Peter230655 replied:
You mean 'b1' instead of b1 or you mena the location of } ?

moorepants replied:
I mean the location of the parentheses and the indentation. Some examples here: https://peps.python.org/pep-0008/#code-lay-out

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,161 @@ def test_attributes_read_only():
+        uy: (-1.0, 1.0),
+        z: (-1.0, 1.0),
+        uz: (-1.0, 1.0),
+    }
```
This case seems to have all unknown variables with bounds. Do you tests cases where bounds are not given for some (or all)?

Peter230655 replied:
> This case seems to have all unknown variables with bounds. Do you tests cases where bounds are not given for some (or all)?

Yes, there is a test without any bounds in the test function. For the state variables, I ignore the bounds.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,161 @@ def test_attributes_read_only():
+    )
+
+    bounds= {
+        a1: (-1.0, 1.0),
```
What would happen if this was? `a1: (-1.0, 1.0)`?

Peter230655 replied:
My method would not care, about IPOPT do not know, never tried.
Would this not be a completely nonstandard way to give an interval?

moorepants replied:
I'm just trying to imagine what inputs may break your function (and thus should be written as tests).

Peter230655 replied:
> I'm just trying to imagine what inputs may break your function (and thus should be written as tests).

Sorry, I mixed this up my PR for ``create_linear_initial_guess``
I did not try with this PR. I will check.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -244,9 +249,58 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
+        if self.bounds is not None:
+            violating_variables = []
+
+            if self.collocator._variable_duration == True:
```
```suggestion
            if self.collocator._variable_duration:
```

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,161 @@ def test_attributes_read_only():
+    with raises(ValueError):
+        _, _ = prob.solve(initial_guess, respect_bounds=True)
+
+    # D: respect_bounds=False by default. Bounds are ignored.
```
you could set max iterations to 1 or 2 to save computation time when solve actually runs.

moorepants commented:
This looks very good. I left a few minor comments, but they aren't necessary to fix for merging.

The only major comment is that you probably need a few more unit tests to check some variation on what type of bounds were set.

Peter230655 commented:
> This looks very good. I left a few minor comments, but they aren't necessary to fix for merging.
> 
> The only major comment is that you probably need a few more unit tests to check some variation on what type of bounds were set.

Thanks!

I just checked an example (your drone) with 'reversed' bounds: {F: (100, -100)}
It did not break, but did not get a reasonable solution at all.
So, if I detect a reversed bound, should I raise an error? If YES, what error? ValueError?

moorepants commented:
I would have thought that IPOPT would raise an error. That makes me wonder why reverse bounds is allowed for IPOPT. Maybe there is a reason you'd want that.

For this PR, we only need to ensure that your new functionality is well tested. Do we want to support reverse bounds with this functionality? If not, we should probably raise an error if reverse bounds are detected. A ValueError sounds correct.

Peter230655 commented:
> I would have thought that IPOPT would raise an error. That makes me wonder why reverse bounds is allowed for IPOPT. Maybe there is a reason you'd want that.
> 
> For this PR, we only need to ensure that your new functionality is well tested. Do we want to support reverse bounds with this functionality? If not, we should probably raise an error if reverse bounds are detected. A ValueError sounds correct.

My method will not work for reversed bounds. (I must admit I did not think of this possibility)
So, I will have to check for this possibility.
At least with opty, I have never come across a situation where I would have needed reversed bounds.
So, maybe better for the opty user to get a ValueError, as it likely will not work, even if IPOPT allows it?

moorepants commented:
>  So, maybe better for the opty user to get a ValueError, as it likely will not work, even if IPOPT allows it?

I'm fine with opty telling you no for this.

Peter230655 commented:
> > So, maybe better for the opty user to get a ValueError, as it likely will not work, even if IPOPT allows it?
> 
> I'm fine with opty telling you no for this.

I will do it like this. Maybe here the numpydocs ``raises`` could be used to explain why the ValueError is raised?

moorepants commented:
Yes, that is the purpose of that documentation section.

Peter230655 commented:
I now check for reversed bounds before I check whether the initial guess meets the bounds. If there is at least one reversed bound a ValueError will be raised.
I test separately for reversed bounds and for violation of the bounds by the initial guess anf for the working of the kwarg in solve.

moorepants commented:
What is the status of this PR? Is it ready for a review?

Looks like merge conflicts need to be resolved.

Peter230655 commented:
> What is the status of this PR? Is it ready for a review?
> 
> Looks like merge conflicts need to be resolved.

Yes, ready.
Let me try and fix the conflicts and I will push again.

Peter230655 commented:
Tried to fix the conflict.

moorepants commented:
There are still outstanding comments from my first review that haven't been addressed.

Peter230655 commented:
- added backend='numpy' to the test function to speed up affairs
- added test if there are no bounds
- simplified the method a little bit
- test that the kwarg in solve works fine

moorepants commented:
Great, but there are still a number of PEP8 related comments I made above that haven't been addressed.

Peter230655 commented:
> Great, but there are still a number of PEP8 related comments I made above that haven't been addressed.

General question regarding test functions:
With this method, and probably more with linear_initial_guess, there is a very large amount of possible combinations. If you test all of them the test function may get very long and slow - if you can think of all of them.
So better to test as many as possible or better to make some 'reasonable' compromise?

Peter230655 commented:
> > So better to test as many as possible or better to make some 'reasonable' compromise?
> 
> You have to be reasonable, but best to have 100% coverage, i.e. test everything.

Thanks! This is what I tried on both: bounds_conflict and linear_initial_guess. (at least what I could think of)

moorepants commented on examples-gallery/advanced/plot_car_in_garage.py:
```diff
@@ -318,17 +318,18 @@ def in_0_1(x):
 }
 
-bounds2 = {py[i]: (0, 100) for i in range(number)}
+bounds2 = {py[i]: (0-epsilon, 100+epsilon) for i in range(number)}
```
```suggestion
bounds2 = {py[i]: (0 - epsilon, 100 + epsilon) for i in range(number)}
```

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -2019,3 +2019,208 @@ def obj_grad(free):
     with raises(ValueError):
         time_vector = prob.time_vector(solution, start_time=t0)
+
+def test_bounds_conflict():
```
I added a `plot=False/True` kwarg to this function so I can check your work. But you have deleted that. Please add it all back and I will review the tests.

moorepants replied:
Sorry, wrong PR.

Peter230655 replied:
> I added a `plot=False/True` kwarg to this function so I can check your work. But you have deleted that. Please add it all back and I will review the tests.

Not intentionally. I tried to merge your changes  - apparently not too successfully so.
Will add this right now!

moorepants commented:
I don't see a test that tests the new kwarg to `.solve()`.

Peter230655 commented:
> I don't see a test that tests the new kwarg to `.solve()`.

![image](https://github.com/user-attachments/assets/1cf84e4e-9ba3-47d3-aba2-0f5c09635394)
I thought, this will test the kwarg?

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -2019,3 +2019,208 @@ def obj_grad(free):
+    # C: respect_bounds=True: initial guess must be within bounds, else a
+    # ValueError is raised.
+    with raises(ValueError):
+        _, _ = prob.solve(initial_guess, respect_bounds=True)
```
Yes, here. Sorry I missed it.

moorepants commented:
I don't see a test where you provide valid bounds and show that it doesn't raise an error. You have an example of not passing any bounds and it doesn't raise, but you need to check when passing bounds.

Peter230655 commented:
> I don't see a test where you provide valid bounds and show that it doesn't raise an error. You have an example of not passing any bounds and it doesn't raise, but you need to check when passing bounds.

It is not there. I will add and push

moorepants commented on opty/direct_collocation.py:
```diff
@@ -244,9 +249,81 @@ def solve(self, free, lagrange=[], zl=[], zu=[]):
+            self.bounds_conflict_initial_guess(free)
         return super().solve(free, lagrange=lagrange, zl=zl, zu=zu)
 
+    def bounds_conflict_initial_guess(self, free):
```
```suggestion
    def check_bounds_conflict(self, free):
```

I suggest this name. It doesn't have to be an initial guess. The action is to check it and raise if there is a conflict.

Peter230655 replied:
> I suggest this name. It doesn't have to be an initial guess. The action is to check it and raise if there is a conflict.

Definitely better! already changed.

moorepants commented:
Ok, this looks good now. I will merge if tests pass. Thanks!

Peter230655 commented:
> Ok, this looks good now. I will merge if tests pass. Thanks!

Thank YOU for your patience!
